modules/layers/src/scatterplot-layer/scatterplot-layer-vertex.glsl.ts

```diff
@@ -31,56 +70,60 @@ attribute vec4 instanceFillColors;
 attribute vec4 instanceLineColors;
 attribute vec3 instancePickingColors;
 
-uniform float opacity;
-uniform float radiusScale;
-uniform float radiusMinPixels;
-uniform float radiusMaxPixels;
-uniform float lineWidthScale;
-uniform float lineWidthMinPixels;
-uniform float lineWidthMaxPixels;
-uniform float stroked;
-uniform bool filled;
-uniform bool antialiasing;
-uniform bool billboard;
-uniform int radiusUnits;
-uniform int lineWidthUnits;
-
 varying vec4 vFillColor;
 varying vec4 vLineColor;
 varying vec2 unitPosition;
 varying float innerUnitRadius;
 varying float outerRadiusPixels;
 
+// Needs to be identical to fragment shader uniforms
```

Comment:
As it needs to be identical, could you define in one place? Derive it from `scatterplot.uniformTypes`?

Reply:
Yes, though there are some tradeoffs to it. So far it seems the shader linker complains loudly if they are different so it does not seem to be a subtle source of errors.

I tried breaking the uniforms into a separate string and injecting it. A minor problem with that  is that I use the vscode GLSL syntax highlighter to make these `glsl` tagged strings readable, and it breaks if I add template string arguments.

I actually wrotecode for generating uniform declarations from uniform types (in the shadertools module), but I haven't started using that yet.

The high-level question for the TSC is perhaps: how much shader code should be autogenerated? Once one starts down that dark path...

Reply:
Do *all* the uniforms need to be typed out in both the vertex and fragment shader? It seems unnecessary to have to include uniforms in the vertex shader that only the fragment shader reads and vice-versa

I do agree there is a tradeoff between saving time and having shaders that are actually readable in the source

Reply:
> Do all the uniforms need to be typed out in both the vertex and fragment shader? It seems unnecessary to have to include uniforms in the vertex shader that only the fragment shader reads and vice-versa

Yes I believe so, though more experience may show otherwise. Also we need to see how this works in both GLSL and WGSL.

Reply:
Agree with you there. Auto-generation does add a layer of complexity, making it all the more important to have solid debugging tools. I'm all for it, just think we really need to keep an eye on the debugging side of things to make sure everything runs smoothly.

Reply:
OK, in the spirit of keeping it simple & landing v9 I'm for just typing out the uniforms in full at this stage